### PR TITLE
Fix Delete Case Type/Prop Errors for Data Dictionary

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

When a new case property is created and, before clicking "Save", a user can now click on the delete icon to remove the case property and have it not be persisted in the back-end.

Additionally, users are now no longer able to delete a case type if it contains a case property being used as the gelocation case property by the geospatial feature.

## Technical Summary

Link to tickets:
- [Prevent case type delete if has geolocation case property](https://dimagi.atlassian.net/browse/SC-3333)
- [Fix error when creating and deleting case property](https://dimagi.atlassian.net/browse/SC-3335)

This PR addresses two bugs related to the feature for deleting case types/props in the Data Dictionary. These fixes will be merged into the feature branch which, in turn, will get merged to master.

## Feature Flag

None

### Safety story

- Local testing
- Automated tests exist

### Automated test coverage

Automated test exists for back-end view to fetch DD case types and properties.

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

This PR will be merged into the `ze/delete-in-dd` branch, so that PR should be reverted.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/data_dictionary/views.py [hunk 1]
```diff
@@ -87,19 +88,20 @@ def data_dictionary_json(request, domain, case_type_name=None):
...
+                    'id': prop.id,
                     'description': prop.description,
                     'label': prop.label,
                     'fhir_resource_prop_path': fhir_resource_prop_by_case_prop.get(
                         prop
                     ),
                     'name': prop.name,
                     'deprecated': prop.deprecated,
-                    'contains_case_data': prop.name in used_props,
+                    'is_safe_to_delete': prop.name not in used_props and prop.name != geo_case_prop,
```
Comment (SmittieC) [inline corehq/apps/data_dictionary/views.py:104]: Thinking about someone in the future seeing this code here: Might they not wonder why the geo_case_property is excluded? I'm wondering if we should put this property in a list of properties that must not be deleted, but I'm not sold on this yet. Thoughts?

Comment (Charl1996) [reply corehq/apps/data_dictionary/views.py:104]: I share this sentiment of not having a seemingly random additional check.

Comment (zandre-eng) [reply corehq/apps/data_dictionary/views.py:104]: That's a valid point. No arguments against putting it in a list, however I think the problem would still remain of future devs wondering why the properties in the list can't be deleted. @SmittieC What do you think about addressing this with a comment or in the documentation itself rather?

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 2]
```diff
@@ -170,6 +171,10 @@ hqDefine("data_dictionary/js/data_dictionary", [
         });
 
         self.confirmDeleteProperty = function () {
+            if (!self.id) {
```
Comment (SmittieC) [inline corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:174]: Just raising the fact that we wouldn't ask users if they are sure they want to delete the property. Personally I think we should still ask users if they are sure, since it's what we do with other properties, so the workflow is the same (we don't surprise users). We could leave out the part mentioning that there's no associated cases/data with this property (or leave it in, because it's true). Curious what others think.

**Review by SmittieC — COMMENTED** [review]
Looks good to me, just two comments

File: corehq/apps/data_dictionary/views.py (hunk 1 above)
Comment (SmittieC) [reply corehq/apps/data_dictionary/views.py:104]: With a comment yes, in docs, no. Reason being that dev's wouldn't read the docs necessarily.

Comment (zandre-eng) [reply corehq/apps/data_dictionary/views.py:104]: >  Reason being that dev's wouldn't read the docs necessarily.

That is true. The other benefit of a code comment would be that devs can get immediate clarity on what's going on instead of having to go digging through documentation.

Given the above, I'll add in a comment.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 3]
```diff
@@ -219,6 +224,10 @@ hqDefine("data_dictionary/js/data_dictionary", [
                     }
 
                     _.each(group.properties(), function (element, index) {
+                        if (element.deleted && !element.id) {
```
Comment (mkangia) [inline corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: I believe if we mark even new ones as deleted like suggested [here](https://github.com/dimagi/commcare-hq/pull/34073/files#r1474504392) we would not need to add `id` at all and just check for `deleted`.

**Review by mkangia — CHANGES_REQUESTED** [review]
Thanks for the PR @zandre-eng 
Left a few suggestions that could make this simpler.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 4]
```diff
@@ -43,13 +43,13 @@ hqDefine("data_dictionary/js/data_dictionary", [
 
                 for (let prop of group.properties) {
                     const isGeoCaseProp = (self.geoCaseProp === prop.name && prop.data_type === 'gps');
-                    if (prop.contains_case_data) {
+                    if (!prop.is_safe_to_delete) {
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:46]: This would be the same, but I wanted to make it explicit that we are marking the case type as unsafe to delete. This conditional check is also required as without it the last looped case property would get to determine whether a case type is safe to delete or not.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: We would still need to check for `id` here since we do not want to try and persist unsaved new case properties which have been immediately deleted (has no id).

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:174]: Comparing the two options, I agree that it might be better to keep things consistent and show a warning for both scenarios. I would opt to not conditionally remove parts of the modal however, especially if the text remains technically true for both cases. This can just help keep things more consistent.

@mkangia I'm curious to hear your thoughts on the recommended approach since you have also reviewed this code?

File: corehq/apps/data_dictionary/views.py [hunk 5]
```diff
@@ -87,7 +87,7 @@ def data_dictionary_json(request, domain, case_type_name=None):
             "is_deprecated": case_type.is_deprecated,
             "module_count": module_count,
             "properties": [],
-            "contains_case_data": len(used_props) > 0,
+            "is_safe_to_delete": len(used_props) == 0,
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/views.py]: [04452c3](https://github.com/dimagi/commcare-hq/pull/34073/commits/04452c314f95be10e53e1e7665826f935db89311)

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:46]: > last looped case property would get to determine whether a case type is safe to delete or not.

Aha. I didn't notice that.
I guess its nicer to just extract that of the loop then to avoid confusion and use
`self.canDelete = !_.every(group.properties, function(prop) { return prop.is_safe_to_delete} )`

I understand if you wish to avoid going through the properties again though I believe the advantage of extracting this out are good to consider.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 2 above)
Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:174]: I'd vote for the option that avoids adding code that isn't worth it.

How bad is it if the user sees the modal for a new property?

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 3 above)
Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: won't these new ones have deleted marked as true if we do what the other comment is saying?

Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: It is true that both new and existing case properties will have `deleted` marked as true. What makes it challenging however, is that there is no way to differentiate which case properties are persisted in the DB and which aren't (as we don't want to send non-persisted deleted props to the back-end). To avoid doing this, we use the ID to check which props exist in the DB and simply don't send deleted props that haven't been saved to the DB yet.

TL:DR - Checking the ID helps determine which delete requests should be sent to the back-end.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:46]: For performance reasons, it's good to try and loop through the array of case properties as few times as possible. This is largely because the entire DD gets loaded on the initial page load and so if there exists a lot of case properties across multiple case types this could further hamper the loading performance of the DD. I think if we were able to incrementally load the DD then this would be much less of a problem.

@mkangia Would a comment here help to remove the confusion? Alternatively, we could also implement something as follows:
```
let unsafePropCount = 0;
for (let group of groupData) {
    ...
    for (let prop of group.properties) {
        if (!prop.is_safe_to_delete) {
            unsafePropCount += 1;
        }
        ...
    }
}
self.canDelete = Boolean(unsafePropCount);
```

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:174]: > How bad is it if the user sees the modal for a new property?

It wouldn't be bad as a user would simply receive a confirmation popup stating that the property is safe to delete. Apart from keeping things consistent, this should also help prevent accidental deletions.

Given the above, I'll remove the code that hides the modal for newly created case properties.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 3 above)
Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: Correct me if I am wrong,
any deleted case property is not sent back to HQ on save?
I thought this was true and so any new or existing case property won't be sent to HQ if it was marked as deleted.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:46]: okay makes sense. 
Good to avoid the reiteration.

Oh, that count change looks an overkill.

How about
```
if (!prop.is_safe_to_delete && self.self.canDelete) {
    self.canDelete = false;
}
```
or else a comment is fine as well.
Thank you!

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:227]: > any deleted case property is not sent back to HQ on save?

We do send deleted case properties back to HQ on save. I think the important thing to note is that case properties are simply marked for delete. Since already saved case properties are saved in the DB, we need to send these case properties that are marked for delete back to HQ so that it can remove these properties from the DB.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:46]: That can work quite nicely. I'll implement the suggested code change, thanks!

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js]: [12423cb](https://github.com/dimagi/commcare-hq/pull/34073/commits/12423cb2b07625b09c52b8f808df582540e1f084)

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js]: [406320d](https://github.com/dimagi/commcare-hq/pull/34073/commits/406320db66b4811c882aa86d79c3323f9b358ff5)

**Review by SmittieC — APPROVED** [review]
Coolios

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:223]: Before the ID check we would send all case properties marked for delete to the back-end, regardless of whether it actually exists in the DB or not. This resulted in the back-end trying to delete a case property that doesn't exist and ultimately causing an exception to be returned to the UI.

Comment (mkangia) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:223]: Apologies, I meant for the case properties that were present on HQ. How does HQ know which one to delete?

Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:223]: For case properties we have an already existing function for updating the case prop with any new changes. This function was extended to check whether the `deleted` flag is true and, if so, remove the case prop instead of updating it.

For more context, you can have a look at [this commit](https://github.com/dimagi/commcare-hq/pull/33956/commits/ace35f75811b69f1daffaa8c6d57aebf95d0e2a0) which adds in the back-end implementation for deleting case types/props.

**Review by mkangia — APPROVED** [review]
